### PR TITLE
Update the GraphProto for subgraphs when saving the Graph. 

### PR DESCRIPTION
**Pull request by skottmckay**
This is required to produce a valid overall Graph if the Graph has been optimized. If the GraphProto attribute are not updated
  a) the optimizations will be lost from any subgraphs
  b) implicit inputs to the subgraphs may be broken by changes to nodes in the parent graph

### CODE REVIEW

File: onnxruntime/core/graph/graph.cc [hunk 1]
```diff
@@ -316,23 +316,31 @@ void Node::SetExecutionProviderType(ProviderType execution_provider_type) {
...
+    proto.set_doc_string(description_);
 
   // Set attributes.
   proto.clear_attribute();
   for (const auto& attribute : attributes_) {
     const gsl::not_null<AttributeProto*> attr{proto.add_attribute()};
     *attr = attribute.second;
-  }
 
+    if (update_subgraphs && attribute.second.has_g()) {
```
Comment (kkaranasos) [inline onnxruntime/core/graph/graph.cc]: Nit: you can use attr instead of attribute.second here.

File: onnxruntime/core/graph/graph.cc [hunk 2]
```diff
@@ -316,23 +316,31 @@ void Node::SetExecutionProviderType(ProviderType execution_provider_type) {
...
+      // partitioning or optimization may have changed the subgraph, so the original GraphProto may not be valid.
+      // call ToGraphProto to get the current info and use this for the attribute
+      const auto& latest = attr_to_subgraph_map_.find(attribute.first)->second->ToGraphProto();
```
Comment (kkaranasos) [inline onnxruntime/core/graph/graph.cc]: Can we know when the subgraph proto needs re-building, so that we avoid it otherwise?

Comment (skottmckay) [reply onnxruntime/core/graph/graph.cc]: ToGraphProto only rebuilds if it has to.

File: onnxruntime/core/graph/graph.cc (hunk 1 above)
Comment (skottmckay) [reply onnxruntime/core/graph/graph.cc]: I could, but I thought it was slightly clearer this way as we're determining if we need to copy things from the existing attribute (attribute.second) into the new one (attr).

File: include/onnxruntime/core/graph/graph.h [hunk 3]
```diff
@@ -298,8 +298,12 @@ class Node {
...
   void SetExecutionProviderType(ProviderType execution_provider_type);
 
-  /** Gets the NodeProto representation of this Node. */
-  void ToProto(ONNX_NAMESPACE::NodeProto& proto) const;
+  /** Gets the NodeProto representation of this Node. 
+  @param update_subgraphs Update the GraphProto values for any subgraphs in the returned NodeProto.
+                          If graph optimization has been run this is most likely required
+                          to ensure the complete Graph is valid.
+  */
+  void ToProto(ONNX_NAMESPACE::NodeProto& proto, bool update_subgraphs = false) const;
```
Comment (KeDengMS) [inline include/onnxruntime/core/graph/graph.h]: >false [](start = 73, length = 5)

where will it be true?

**Pull request merged**